Allow `jl_set_const` to overwrite const global (with warning).

NHDaly opened the pull request:
Currently, `jl_set_const` (and `jl_set_global`) silently do
nothing if the binding is a `const` global variable. This does not match
the `julia` behavior, which is to allow redefining the const global
(with a Warning) as long as *the types are the same*.

I've added tests for those functions, and include two broken tests demonstrating the above.

I'm also going to push up a commit that changes those two functions to have the
same behavior as an assignment to a const global in julia, by simply delegating
to `jl_checked_assignment`.

NHDaly committed:
Allow `jl_set_const` to overwrite const global (with warning).

Before this commit, `jl_set_const` (and `jl_set_global`) silently do
nothing if the binding is a `const` global variable. This does not match
the `julia` behavior, which is to allow redefining the const global
(with a Warning) as long as *the types are the same*.

This commit changes those two functions to have the same behavior, by
simply delegating to `jl_checked_assignment`.

NHDaly committed:
Add tests for jl_set_[global/const]: expose bug preventing overwrite.

Adds a broken test showing that you currently cannot overwrite a const
global via `jl_set_const`, even if the type is unchanged.

JeffBezanson commented:
It might make sense for these to just throw an error in the case where they previously did nothing. That fits internal functions a bit better.

NHDaly commented:
Yeah that could make sense as well.

However, rather than (or in addition to?) throwing an error, it would be
nice to signal an error back to the c caller, so that they can respond
accordingly.


----

And on that note, if we did change these methods to simply error, is there
an available API that the `C` caller can use in order to overwrite the
value? Should I use jl_checked_assignment directly instead?

Thanks!



On Sun, Jun 10, 2018, 11:15 AM Jeff Bezanson <notifications@github.com>
wrote:

> It might make sense for these to just throw an error in the case where
> they previously did nothing. That fits internal functions a bit better.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/JuliaLang/julia/pull/27520#issuecomment-396056969>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABgkEYYHOP1yvQsvww1Z1BG98OOOGOEGks5t7TgGgaJpZM4UhlCe>
> .
>

NHDaly commented:
No, I haven't done anything since the last message.

There is still an open question:
1. Can we return an error to the C caller instead of throwing a julia exception? That way the C caller can know to *get* the binding and change its value directly.
 But if that's not desirable I can just make the change to throw an error...

NHDaly commented:
Also, I guess the other question is still:

2. are we sure we don't want to allow redefining an existing const? It seems weird to me that this doesn't match standard julia behavior. Otherwise, I think the user would have to write something like this in order to avoid an exception:

```
if (jl_is_const(m, var)) {
    jl_checked_assignment(jl_get_binding_wr(m, var, 1), val);
} else {
    jl_set_const(m, var, val);
}
```
Instead of simply calling `jl_set_const`.

NHDaly commented:
------
Also, actually, one more thing (sorry):

Looking at that above example just made me realize the C api allows you to change a variable from non-const to const, which actually seems like a bug? I'm opening a new issue to cover that one:
https://github.com/JuliaLang/julia/issues/28106

NHDaly committed:
Emit error in jl_set_const/global if already const.

Put back the old behavior for jl_set_const,jl_set_global, which is that
you cannot overwrite an existing global const. Added a julia error
instead of the previous silent failure.

NHDaly commented:
Also, i've uploaded the requested fix so there's something concrete to discuss! :)

JeffBezanson commented:
On revisiting this, it doesn't seem great for all ways to set variables from the C API to involve printing messages and/or throwing errors --- not very C-like. Most other functions like this simply assume (and `assert`) that the arguments are valid (e.g. `jl_get_nth_field` asserts that the index is in bounds). So I would be fine with removing the checks and replacing them with assertions. Or, if there is actually a use case for finding out whether the operation succeeded, we could return an error code.

NHDaly commented:
Yeah that makes sense to me.

I guess, my only other comment before proceeding is that the name `jl_set_const` implies to me that it's providing the C-equivalent to the following julia functionality:
```julia
const var = val
```
So I would expect it to behave the same as that (ie allow changing value, but not changing type).

If we are saying that this method is _really_ "create a brand-new `const` global iff one doesn't already exist," can we rename it to reflect that before adding assertions? 

Perhaps `jl_new_const` to match `jl_new_array` and `jl_new_module`?
The other functions that currently start with `jl_set_` are `jl_set_ARGS`, `jl_set_nth_field`, `jl_set_typeof`, and `jl_set_errno`, which mostly all modify an existing thing?

If renamed like that, i think just having asserts would be totally fine! :) What do you think?

NHDaly commented:
> Or, if there is actually a use case for finding out whether the operation succeeded, we could return an error code.

Yeah actually maybe there isn't. In my use-case, I was trying to set the global `ARGS` when compiling a program, but I was surprised that `jl_set_const` wouldn't let me overwrite it. But then I realized it would be easier to just get the existing array and fill it:
https://github.com/JuliaLang/PackageCompiler.jl/commit/900551b5311953c98d588871be5357fd5118d044#diff-6b3a47e708d881362b8f7510db82c15cR41

So I do not have a use-case that needs a returned status code, and so i agree, until then it seems reasonable to just error out.

JeffBezanson commented:
`new` usually implies allocating and returning an object, so maybe `jl_define_const`?

NHDaly commented:
ohh good point. Yeah I like that. Thanks for bikeshedding with me.

So to review:
- `jl_define_const` will create a new global const with the supplied value, and error (or assert) if the variable already exists.

What about `jl_set_global` then? Currently we're defining the behavior here to be:
- create a new non-const global if the variable name doesn't exist, or change it to a new value if it does already exist. Raise an error (or assert) if the variable is already a const global.

Are we okay leaving it like that? I think that sounds good to me!

NHDaly committed:
s/jl_set_const/jl_define_const/g && update behaviors:

 - jl_define_const will create a new global const with the supplied
      value, and error (or assert) if the variable already exists.
 - jl_set_global will create a new non-const global if the variable name
      doesn't exist, or change it to a new value if it does already exist,
      raising an error if the variable is already a const global.

NHDaly committed:
Change error() to assert in jl_set_global/jl_define_const.

NHDaly commented:
Hi, sorry for the delay on this!

So, I've made the changes we talked about. I've changed the name to `jl_define_const` (hopefully gotten them all) and i've changed their behaviors as we discussed.

That's as of 50ed58f7a83b8f45a8d4fa498bb88ee13365d3b2. They're still emitting errors, though.

I wasn't super sure how to test the `assert`s. I'm also uploading one more commit, 0b416c51b652830b1d47b94af2396dd44fecda88, where I changed the errors to asserts, but i wasn't sure how to test it. My gut tells me you just _don't_ test it, since it's part of the function's preconditions. But we don't _document_ that precondition anywhere. Can I add comments about it to `julia.h`? I know there's not a lot of documentation there, but is it okay to add to it?

thanks! :)

NHDaly commented:
(Or, we can just leave them as errors. I'm fine with whatever so it's up to whatever you think is best! 😄)

NHDaly committed:
Add comments to julia.h documenting preconditions

Add comments to both jl_set_global and jl_define_const documenting their effects and preconditions.

NHDaly committed:
Fix jl_define_const: Fixed to _actually_ check for existing vars.

JeffBezanson commented on src/module.c:
```diff
@@ -482,20 +482,21 @@ JL_DLLEXPORT jl_value_t *jl_get_global(jl_module_t *m, jl_sym_t *var)
-        jl_gc_wb(m, val);
-    }
+    // Assert that it doesn't already exist.
+    jl_binding_t *bp = jl_get_binding(m, var);
```
This won't work, since it will cause the binding to be resolved (possibly to something visible via `using`), when we want to create a new variable owned by this module. The binding also doesn't have to be totally absent; other operations might have added a binding object to the table without setting a value for the variable.

NHDaly replied:
Ah, okay, thanks. Can i just check if the `value` is `NULL`, then? Is a `NULL` `jl_binding_t.value` different than a variable whose value is `nothing`? I'm not that familiar with the C api still, sorry.

NHDaly replied:
(Also, sorry i haven't been testing these commits before pushing.. I'm having trouble getting my local julia to build for some reason. i might need to blow it away and start over.)

JeffBezanson commented on src/julia.h:
```diff
@@ -1325,10 +1325,12 @@ JL_DLLEXPORT int jl_defines_or_exports_p(jl_module_t *m, jl_sym_t *var);
-JL_DLLEXPORT void jl_set_const(jl_module_t *m JL_ROOTING_ARGUMENT,
+// Define a new const global variable. NOTE: var must not already exist.
+JL_DLLEXPORT void jl_define_const(jl_module_t *m JL_ROOTING_ARGUMENT,
                                jl_sym_t *var,
```
indent

NHDaly replied:
Fixed. thanks.

NHDaly committed:
Fix indent in julia.h after rename

NHDaly commented:
Hi all - i'm sorry i let this PR languish. I'm going to close it, since it's quite out of date. If someone feels they want to pick this up, please feel free to!! <3
